cmd/clusterctl/client/cluster/cert_manager.go

```diff
@@ -220,86 +228,90 @@ func (cm *certManagerClient) createObj(o unstructured.Unstructured) error {
-	webhook, err := cm.getWebhook()
-	if err != nil {
-		return false, errors.Wrap(err, "failed to check if the cert-manager web-hook exists")
-	}
-	if webhook != nil {
-		return true, nil
-	}
-	return false, nil
-}
+// waitForAPIReady will attempt to create the cert-manager 'test assets' (i.e. a basic
+// Issuer and Certificate).
+// This ensures that the Kubernetes apiserver is ready to serve resources within the
+// cert-manager API group.
+// If retry is true, the createObj call will be retried if it fails. Otherwise, the
+// 'create' operations will only be attempted once.
+func (cm *certManagerClient) waitForAPIReady(ctx context.Context, retry bool) error {
+	log := logf.Log
+	// Waits for for the cert-manager web-hook to be available.
+	log.Info("Waiting for cert-manager to be available...")
 
-// isWebhookAvailable returns true if the cert-manager web-hook has the condition type:Available with status:True.
-// This is required to check the web-hook is working and ready to accept requests.
-func (cm *certManagerClient) isWebhookAvailable(webhook *unstructured.Unstructured) (bool, error) {
-	conditions, found, err := unstructured.NestedSlice(webhook.Object, "status", "conditions")
+	testObjs, err := getTestResourcesManifestObjs()
 	if err != nil {
-		return false, errors.Wrap(err, "invalid cert-manager web-hook: failed to get conditions")
-	}
-
-	// if status.conditions does not exists, we assume the web-hook is still starting
-	if !found {
-		return false, nil
+		return err
 	}
 
-	// look for the condition with type:Available and status:True or return false
-	for _, condition := range conditions {
-		conditionMap, ok := condition.(map[string]interface{})
-		if !ok {
-			return false, errors.Wrap(err, "invalid cert-manager web-hook: failed to parse conditions")
-		}
+	for i := range testObjs {
+		o := testObjs[i]
+		log.V(5).Info("Creating", logf.UnstructuredToValues(o)...)
 
-		conditionType, ok := conditionMap["type"]
-		if !ok {
-			return false, errors.Wrap(err, "invalid cert-manager web-hook: there are conditions without the type field")
+		// Create the Kubernetes object.
+		// This is wrapped with a retry as the cert-manager API may not be available
+		// yet, so we need to keep retrying until it is.
+		if err := cm.pollImmediateWaiter(waitCertManagerInterval, cm.getWaitTimeout(), func() (bool, error) {
+			if err := cm.createObj(o); err != nil {
+				log.V(5).Info("Failed to create cert-manager test resource", logf.UnstructuredToValues(o)...)
+
+				// If retrying is disabled, return the error here.
+				if !retry {
+					return false, err
+				}
+
+				return false, nil
+			}
+			return true, nil
+		}); err != nil {
+			return err
 		}
-
-		if conditionType != "Available" {
-			continue
+	}
+	for i := range testObjs {
+		obj := testObjs[i]
+		if err := cm.deleteObj(obj); err != nil {
+			log.V(5).Info("Failed to delete cert-manager test resource", logf.UnstructuredToValues(obj)...)
+			return err
```

Comment:
I feel if there may be some interesting behavior here. If we delete the namespace `cert-manager-test` and then delete the certificate/issuer could we run into a chance of getting a NotFound error since deleting the namespace would delete the stuff inside it?

I haven't run into it while testing manually...but I'm wondering if we should ignore NotFound errors here. WDYT?

Reply:
Might be we can simplify and simply delete the ns...

Reply:
I've adjusted this to tolerate NotFound errors instead for now, given there is no specific namespace creation logic this seems simpler than having to inspect the manifests and infer what namespace to delete. This are can be improved as you've noted above to randomise the namespace name, which I think will make deletion a little simpler then too 😄